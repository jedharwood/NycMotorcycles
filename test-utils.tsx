import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import en from './languages/en.json';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import { QueryClient, QueryClientProvider } from 'react-query';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  const queryClient = new QueryClient();

  return (
    <MemoryRouterProvider>
      <QueryClientProvider client={queryClient}>
        <IntlProvider locale='en' messages={en}>
          {children}
        </IntlProvider>
      </QueryClientProvider>
    </MemoryRouterProvider>
  );
};

const renderWithProviders = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { renderWithProviders as render };
