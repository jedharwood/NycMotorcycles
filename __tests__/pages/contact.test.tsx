import ContactPage from '@/pages/contact';
import { render, screen, waitFor } from '@/test-utils';
import userEvent from '@testing-library/user-event';

const findAndPopulateInputAsync = async (screen: any, testId: string, value: string): Promise<void> => {
    const input: HTMLElement = screen.getByTestId(testId);
    await userEvent.type(input, value);
}

const emailValidationMessage: string = 'Please enter a valid email address';
const requiredFieldValidationMessage: string = 'This is a required field';

describe('ContactPage', () => {
  it('should render page in English', () => {
    const { container } = render(<ContactPage />);

    expect(container).toMatchSnapshot();
  });

  it('validator should display message for invalid email input', async () => {
    render(<ContactPage />);
    await findAndPopulateInputAsync(screen, 'email', 'invalid email address');
    await findAndPopulateInputAsync(screen, 'senderName', 'valid input');
    await findAndPopulateInputAsync(screen, 'subject', 'valid input');
    await findAndPopulateInputAsync(screen, 'message', 'valid input');

    const submitButton: HTMLElement = screen.getByTestId('contact-form-submit-button');

    await waitFor(() => {
        userEvent.click(submitButton);
        expect(screen.queryByText(emailValidationMessage)).toBeInTheDocument();
        expect(screen.queryByText(requiredFieldValidationMessage)).not.toBeInTheDocument();
    });

    // Assert that post has not been called
  });
});


