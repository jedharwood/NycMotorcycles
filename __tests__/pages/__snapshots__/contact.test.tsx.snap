// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactPage should render page in English 1`] = `
<div>
  <main
    class="w-full rounded-md bg-stone-600 bg-opacity-90 py-4 px-6 text-stone-50 shadow-lg"
  >
    <div
      class="mb-4 text-center"
    >
      <h2
        class="text-xl font-medium opacity-80 md:text-2xl"
      >
        Contact us
      </h2>
      <p>
        Lets talk about bikes
      </p>
    </div>
    <form
      class="space-y-6"
      novalidate=""
    >
      <div
        class="flex justify-center"
      >
        <div
          class="w-full md:w-3/4 lg:w-1/2"
        >
          <label
            class="mb-2 block"
            for="email"
          >
            Your email
          </label>
          <input
            class="bg-stone-50 bg-opacity-90 rounded-md w-full p-2.5 text-gray-900 outline-none focus:ring focus:ring-offset hover:ring hover:ring-offset focus:ring-teal-500 hover:ring-teal-500"
            data-testid="email"
            name="email"
            placeholder="name@domain.com"
            type="text"
          />
        </div>
      </div>
      <div
        class="flex justify-center"
      >
        <div
          class="w-full md:w-3/4 lg:w-1/2"
        >
          <label
            class="mb-2 block"
            for="senderName"
          >
            Your name
          </label>
          <input
            class="bg-stone-50 bg-opacity-90 rounded-md w-full p-2.5 text-gray-900 outline-none focus:ring focus:ring-offset hover:ring hover:ring-offset focus:ring-teal-500 hover:ring-teal-500"
            data-testid="senderName"
            name="senderName"
            placeholder="Indian Larry"
            type="text"
          />
        </div>
      </div>
      <div
        class="flex justify-center"
      >
        <div
          class="w-full md:w-3/4 lg:w-1/2"
        >
          <label
            class="mb-2 block"
            for="subject"
          >
            Subject
          </label>
          <input
            class="bg-stone-50 bg-opacity-90 rounded-md w-full p-2.5 text-gray-900 outline-none focus:ring focus:ring-offset hover:ring hover:ring-offset focus:ring-teal-500 hover:ring-teal-500"
            data-testid="subject"
            name="subject"
            placeholder="How can we help you"
            type="text"
          />
        </div>
      </div>
      <div
        class="flex justify-center"
      >
        <div
          class="w-full md:w-3/4 lg:w-1/2"
        >
          <label
            class="mb-2 block"
            for="message"
          >
            Your message
          </label>
          <textarea
            class="bg-stone-50 bg-opacity-90 rounded-md w-full p-2.5 text-gray-900 outline-none focus:ring focus:ring-offset hover:ring hover:ring-offset focus:ring-teal-500 hover:ring-teal-500"
            data-testid="message"
            name="message"
            placeholder="Leave a comment..."
            rows="6"
          />
        </div>
      </div>
      <div
        class="flex justify-center"
      >
        <button
          class="rounded-md px-6 py-2 text-lg bg-teal-700 hover:bg-teal-500"
          data-testid="contact-form-submit-button"
          type="submit"
        >
          Send
        </button>
      </div>
    </form>
  </main>
</div>
`;
